addons/stock/models/stock_move.py

```diff
@@ -2159,14 +2159,34 @@ def _trigger_assign(self):
                                                          order='reservation_date, priority desc, date asc, id asc')
         moves_to_reserve._action_assign()
 
+    def _rollup_move_dests_fetch(self):
```

Comment:
These two methods are nice although I'm wondering, do we need new methods for this ? Can't these be merged into the existing `_rollup_move_xxx` ones ? Because the current ones are only going down the move_origs/dests tree so it looks like they could be merged. Also, splitting the fetching means that we need to explicitly call the new methods before so it's slightly more error-prone.

Reply:
I don't think there is a way to merge the fetching logic directly into `_rollup_move_xxx` methods because they are only ever called one move at a time.
Like this for outs: https://github.com/odoo/odoo/blob/c02406b90822b8715df96d82747ee23dd4834fc6/addons/stock/report/stock_forecasted.py#L278-L280
and this for ins:  https://github.com/odoo/odoo/blob/c02406b90822b8715df96d82747ee23dd4834fc6/addons/stock/report/stock_forecasted.py#L303-L308

You can't just call `outs._rollup_move_origs()` because the resulting ordered set isn't the same as calling one at a time in a loop over outs. (Because doing it on all outs would only return a single ordered set of all the moves in the tree, but doing each one at a time only returns the set for that specific move.)

Though I do agree that it is messy to keep a separate method for doing this. Now that I think more about it, you could probably achieve the same thing that this does: https://github.com/odoo/odoo/blob/c02406b90822b8715df96d82747ee23dd4834fc6/addons/stock/report/stock_forecasted.py#L272-L274

By just calling the original method on outs/ins but ignore the result:
```py
        outs._rollup_move_origs()
        ins._rollup_move_dests()
```
Haven't tested but I think that would acheive the same purpose of "pre-fetching" all the nodes in an efficient manner without duplicating the logic in the new fetch methods.

If littering the pre-fetching logic in the code is a concern, you could possibly introduce instead something like `_rollup_all_move_xxx` which would combine the pre-fetching logic with the for loop logic like so:
```py
def _rollup_all_move_origs(self):
    # Pre-fetch all moves
    self._rollup_move_origs() # or self._rollup_move_origs_fetch()?
    return {
        move: move._rollup_move_origs()
        for move in self
    }
```

Then the other code could be re-worked to use this without much difficulty, and it would "hide" the concerned logic.

Reply:
I'm wondering. 89 vs 3 is still an improvement but it's not that much of a difference given the huge nomber of moves to process in this specific database. I would say that the existing methods are enough but I'll leave that up to the logistics team.

Reply:
It may also be possible to change `_rollup_move_xxx` use `read` instead of dot access, and that *might* work for fetching more than 1000 at a time. I'm not sure if there could be any issues from NewIds by doing that though.

Reply:
^ Yes so making this change in `_rollup_move_xxx` to use `read` instead:
```diff
diff --git a/addons/stock/models/stock_move.py b/addons/stock/models/stock_move.py
index cc9e1b46b2e8..9eb65d445dd2 100644
--- a/addons/stock/models/stock_move.py
+++ b/addons/stock/models/stock_move.py
@@ -2186,7 +2186,8 @@ Please change the quantity done or the rounding precision of your unit of measur
         if not unseen:
             return seen
         seen.update(unseen)
-        self.filtered(lambda m: m.id in unseen).move_dest_ids._rollup_move_dests(seen)
+        to_visit_ids = [m['id'] for m in self.filtered(lambda m: m.id in unseen).read(['move_dest_ids'])]
+        self.browse(to_visit_ids)._rollup_move_dests(seen)
         return seen
 
     def _rollup_move_origs(self, seen=False):
@@ -2196,7 +2197,8 @@ Please change the quantity done or the rounding precision of your unit of measur
         if not unseen:
             return seen
         seen.update(unseen)
-        self.filtered(lambda m: m.id in unseen).move_orig_ids._rollup_move_origs(seen)
+        to_visit_ids = [m['id'] for m in self.filtered(lambda m: m.id in unseen).read(['move_orig_ids'])]
+        self.browse(to_visit_ids)._rollup_move_origs(seen)
         return seen
 
     def _get_forecast_availability_outgoing(self, warehouse):

```

Would allow for pre-fetching without limit of 1000. So it could be used instead of the fetch methods. Though am unsure about if this is a breaking change or not.

Reply:
Possible breaking change there if `read` does or doesn't give duplicate IDs for if `unseen`'s `move_xxx_ids` has duplicates or not. Versus if using dot access does or doesn't give duplicate IDs. Will need to test that case.